phoenix-core-client/src/main/java/org/apache/phoenix/schema/CompiledConditionalTTLExpression.java

```diff
@@ -0,0 +1,307 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        CompiledConditionalTTLExpression that = (CompiledConditionalTTLExpression) o;
+        return ttlExpr.equals(that.ttlExpr);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(ttlExpr);
+    }
+
+    @Override
+    public String getTTLExpression() {
+        return ttlExpr;
+    }
+
+    @Override
+    public String toString() {
+        return getTTLExpression();
+    }
+
+    /**
+     * The cells of the row (i.e., result) read from HBase store are lexicographically ordered
+     * for tables using the key part of the cells which includes row, family, qualifier,
+     * timestamp and type. The cells belong of a column are ordered from the latest to
+     * the oldest. The method leverages this ordering and groups the cells into their columns
+     * based on the pair of family name and column qualifier.
+     */
+    private List<Cell> getLatestRowVersion(List<Cell> result) {
+        List<Cell> latestRowVersion = new ArrayList<>();
+        Cell currentColumnCell = null;
+        long maxDeleteFamilyTS = 0;
+        for (Cell cell : result) {
+            if (currentColumnCell == null ||
+                    !CellUtil.matchingColumn(cell, currentColumnCell)) {
+                // found a new column cell which has the latest timestamp
+                currentColumnCell = cell;
+                if (currentColumnCell.getType() == Cell.Type.DeleteFamily ||
+                        currentColumnCell.getType() == Cell.Type.DeleteFamilyVersion) {
+                    // DeleteFamily will be first in the lexicographically ordering because
+                    // it has no qualifier
+                    maxDeleteFamilyTS = currentColumnCell.getTimestamp();
+                    // no need to add the DeleteFamily cell since it can't be part of
+                    // an expression
+                    continue;
+                }
+                if (currentColumnCell.getTimestamp() > maxDeleteFamilyTS) {
+                    // only add the cell if it is not masked by the DeleteFamily
+                    latestRowVersion.add(currentColumnCell);
+                }
+            }
+        }
+        return latestRowVersion;
+    }
+
+    /**
+     * This logic has been borrowed from GlobalIndexChecker#removeOlderCells
+     * This is needed to correctly handle updates to the index row when some columns
+     * are set to null and thus are not updated but an older version of the same column
+     * might have a valid value. Since TTLRegionScanner evaluates the row before GlobalIndexChecker
+     * processes the row we need to mask such columns in TTLRegionScanner also for correctness.
+     * @return
+     */
+    private List<Cell> maskOlderCells(List<Cell> cellList) {
```

Comment:
Instead of borrowing/copying, we can make this method to a utility method in ScanUtil.